### PR TITLE
fix(jsonpath): tests and code for #646 and #651

### PR DESCRIPTION
**Pull request by AlexBHarley**
Courtesy of [@honorcode](https://github.com/honorcode), solves #646 and #651

Pretty happy with this, @yasserf do you want to give it a second pair of eyes.

### CODE REVIEW

File: src/record/json-path.js [hunk 1]
```diff
@@ -51,19 +82,22 @@ JsonPath.prototype.setValue = function (node, value) {
...
  */
 JsonPath.prototype._tokenize = function () {
-  const parts = this._path.split(SPLIT_REG_EXP)
+
+  // makes json path array items a single 'part' value of parts below
+  // 'arrayProp[#]' members transform to 'arrayProp=#' now instead of 'arrayProp.#' previously
+  // see setValue fnc above for special handling of array item parsing vs numeric obj member name
+  // e.g. 'object.1' parsing. this allows for support of parsing and differentiating object
+  // member names that are also numeric values
+  let str = this._path.replace(/\s/g, '')
```
Comment (yasserf) [inline src/record/json-path.js]: this is super expensive, doing three inline regexs and not even compiling them all upfront.

**Comment by yasserf** [comment]
This is a pretty expensive solution.. I would need to look into this more first but I think it can possibly be done in a more performant manner.

**Comment by yasserf** [comment]
I'm guessing this is also a backwards compatability change to how people used the . vs []

This is certainly correctly behaviour, just a bit concerned about impact it may cause merging it in its current state.

**Comment by AlexBHarley** [comment]
So discussed this with @yasserf a bit more. We don't want to do 3 regexes here

```
let str = this._path.replace(/\s/g, '')
str = str.replace(/\[(.*?)\]/g, '=$1')
const parts = str.split(SPLIT_REG_EXP)
```

The first two should be compiled like the third one, but ideally we'd still only want to do one.

A couple of solutions discussed to avoid the `replace(regex, '=$1)`. 

- We can store some metadata somewhere about token when surrounded by `[` and `]`, this way we know it's an array and not a property
- Seemingly better solution: 
  - when we find a token surrounded by `[` and `]`, put it into the token array explicitly as a number (`Number(token)`)
  - otherwise, put it in as a string (`String(token)`)

This way inside the `setValue` function we can see by the type of the token, whether we should be operating on an array or object.

**Comment by yasserf** [comment]
Closing this as its pretty expensive and needs to be addressed in a different manner.

**Pull request closed without merging**